Add algorithm for creating POLREF transmission workspaces

rbauststfc opened the pull request:
### Description of work

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->

A new algorithm that implements the pre-processing steps described on the linked issue. See the "_**Calibration runs**_" section for a description of what is required. The purpose of the algorithm is to pre-process the magnetic and non-magnetic calibration runs taken on POLREF so that they are ready to pass to the `PolarizationEfficienciesWildes` algorithm.

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->

Fixes #35682. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->
This PR includes a few workarounds in order to prevent delays to the work. However, ideally more fundamental fixes should be found, so I will open issues for the following:

- It seems that you cannot retrieve the output workspace group from an `OutputWorkspace` property when an algorithm is run as a child on a workspace group. As explained in the comment in the code, I have implemented a workaround, but it would be good to try and resolve this at source at some point (#38473).

- The `Divide` algorithm sets the YUnit to an empty string and the distribution flag to `True` if two workspaces with the same units are divided by each other. However the Mantid  `Integration` algorithm performs the integration differently for distribution data as it assumes that it means the counts in the workspace have been divided by the bin width. The apparent discrepancy between the two definitions of "distribution" can lead to incorrect results when we create the transmission workspace after a flood correction (due to the use of the `Integration` algorithm within `CreateTransmissionWorkspaceAuto` in order to normalize by monitors). In this PR, I have manually set the flood corrected workspace back to a non-distribution as a workaround.

### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

This implements a workflow that the POLREF scientists have already implemented in their own code, so I have validated against their results to check that the new algorithm produces equivalent results.

The reviewer should run the new algorithm and check that the validation is robust. The following script can be used to do that:

```
out_grp = ReflectometryISISCreateTransmission(InputRuns=["POLREF32130", "POLREF32132"], ProcessingInstructions="272-290")
```

The algorithm docs should be built and checked to ensure that they are easy to understand, and there are no spelling and grammar issues.

I have run a build from branch to check that the system test passes on all operating systems, although this was a few weeks ago now so could be repeated by the reviewer if they want to.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

rbauststfc committed:
Add new alg ReflectometryISISCreateTransmission

rbauststfc committed:
Add information logging

rbauststfc committed:
Add unit tests for ReflectometryISISCreateTransmission

rbauststfc committed:
Add algorithm documentation

rbauststfc committed:
Add error handling for background subtraction alg

The background subtraction algorithm has a ProcessingInstructions
property, which is the same name as the property on
ReflectometryISISCreateTransmission. This means that error messages can be
confusing if we don't explicitly mention that they have come from the
background subtraction algorithm.

rbauststfc committed:
Change input parameter to take a list of run nums

Use the LoadAndMerge algorithm to load and sum all the runs into a
single workspace as part of the first step.

rbauststfc committed:
Add system test for ReflectometryISISCreateTransmission

rbauststfc committed:
Set ws to non-distribution after flood correction

The flood correction performs a divide which can result in the corrected
workspace being set as a distribution. This affects the result of the integration
that takes place when normalising by monitors during the
transmission workspace creation. The flood corrected workspace should
not be marked as a distribution, according to the definition that the
integration algorithm is using for this, so we need to set the flag
back to False.

rbauststfc committed:
Add release note for PR 38466

adriazalvarez commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCreateTransmission.py:
```diff
@@ -0,0 +1,179 @@
+        # The flood correction performs a divide, which can result in the workspace being set as a distribution
+        # if the flood and input workspaces have the same Y units. We need to set the resulting workspace back to
+        # a non-distribution as it may be dimensionless after the divide, but it shouldn't be a distribution.
+        if isinstance(corrected_ws, WorkspaceGroup):
```
Is there any situation where we want the output of apply flood workspace as a distribution ? If this is the case we could make this change on the apply flood workspace algorithm instead of here.

rbauststfc replied:
That's a good thought and something that we might do at some point in the future, but it would need careful testing as that might change results in other parts of the code that use `ApplyFloodWorkspace`. For now I've been discussing the underlying behaviour in the `Divide` algorithm with Richard and Pete to decide whether we should be providing a way of overriding the default behaviour there. If we did that then we could review some of our workflows more generally to make use of that, but it will be a bigger investigation and out of scope of this PR I think.

adriazalvarez commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCreateTransmission.py:
```diff
@@ -0,0 +1,179 @@
+        def run_alg(ws):
+            args.update({input_ws_prop_name: ws, "OutputWorkspace": ws})
+            alg = self.createChildAlgorithm(alg_name, **args)
+            alg.setRethrows(True)
```
I think if the algorithm is a child algorithm, it will rethrow anyway. So this line may not be necessary.

rbauststfc replied:
I had always understood it to be the opposite actually, that a child algorithm won't rethrow automatically. I might test to double check at some point, but I seem to recall finding that from previous experience, so I'll keep this in if that's OK.

adriazalvarez replied:
I looked it up on the algorithm.cpp class: 
```      
if (m_isChildAlgorithm || m_runningAsync || m_rethrow) 
       throw;
```
Maybe there are other parts of the algorithms that interact with it too, or I don't know if it works differently in data processor algorithms. In any case, as it is is not hurting anyone.

rbauststfc replied:
I found an old comment from Martyn on Slack that said child algorithms would set this automatically. I may have got confused by some behaviour I saw in some of our unit tests quite a while ago. It looks like the Reflectometry code does use `setRethrows` with child algorithms quite often (I noticed it quite a bit in `ReflectometryISISLoadAndProcess`), but I think you're right that it's probably not necessary. As you say, doesn't really do any harm, but one to remember going forward, so thanks for flagging it up!

adriazalvarez commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCreateTransmission.py:
```diff
@@ -0,0 +1,179 @@
+        # If the input run loads as a workspace group, then using the default workspace group handling of other Mantid algorithms
+        # prevents us retrieving an output workspace group from this algorithm when it is run as a child (unless we store things
+        # in the ADS). To avoid this, when we have a workspace group we loop through it, run each algorithm against the child
+        # workspaces individually and then collect them into a group again at the end
```
Do you think that we could reference the issue #38473 in this comment?

adriazalvarez commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCreateTransmission.py:
```diff
@@ -0,0 +1,179 @@
+
+        return corrected_ws
+
+    def _apply_background_subtraction(self, workspace):
```
The algorithm `CreateTransmissionWorkspace` is also subtracting a flat background. What's the difference here? Can this be omitted?

rbauststfc replied:
It's an interesting question, but I don't know whether you would get the same result if you took this out - you don't seem to be able to pass a background region to subtract in the way you can with `ReflectometryBackgroundSubtraction`, for example? I haven't looked too closely at it as the task with this one was to implement the same workflow as POLREF use in their own code, and they use `ReflectometryBackgroundSubtraction` followed by `CreateTransmissionWorkspace`, so I stuck with that.

adriazalvarez commented on Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCreateTransmissionTest.py:
```diff
@@ -0,0 +1,149 @@
+        self._check_output_data(output_ws, 771, 0.018547099002195058, 1.2884763705979813e-05)
+
+    def test_correct_output_for_workspace_groups(self):
+        expected_history = [
```
Sorry if this is obvious . I haven't worked much with the history until now. Why each workspace contained in a group has repeated history?

rbauststfc replied:
I couldn't tell you the underlying reason as to why this happens I'm afraid, it's just the way the history seems to be recorded on the child workspaces of workspace groups. I think we've talked previously about the fact that the history for workspace groups has some issues and that this may be addressed in an epic at some point in the future.

adriazalvarez commented on Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCreateTransmissionTest.py:
```diff
@@ -0,0 +1,149 @@
+        self._check_output_data(output_ws, 771, 0.018547099002195058, 1.2884763705979813e-05)
+
+    def test_correct_output_for_workspace_groups(self):
+        expected_history = [
+            self._LOAD_ALG,
+            self._FLOOD_ALG,
+            self._FLOOD_ALG,
+            self._BACK_SUB_ALG,
+            self._BACK_SUB_ALG,
+            self._TRANS_WS_ALG,
+            self._TRANS_WS_ALG,
+        ]
```
I think it is possible to refactor the history recreation list in a more compact form (similar in other lists of the file):
```suggestion
        expected_history = [self._LOAD_ALG, *[self._FLOOD_ALG]*2,                                          *[self._BACK_SUB_ALG]*2, *[self._TRANS_WS_ALG]*2]

```
but since the list is not that long, and it is clear enough as it is. I leave this change up to you.

rbauststfc replied:
That's interesting syntax that I don't think I've noticed before, thanks for mentioning. It looks like you'd be creating a series of lists this way that all end up in one? I think, as you say, because this list is not that long I might leave it as it is because I think it's currently quite straightforward to see what's going on. Handy to know about this for longer lists, though, as I can definitely see there would be an advantage there.

adriazalvarez reviewed (changes requested):
Thanks for adding this algorithm. I have followed the test instructions and the transmission workspace is created with the intended properties. Additionally, unit tests and system tests are passing. The docs are well formatted and clear to follow. 
I have few -minor- suggestions and questions, but otherwise it looks good.

rbauststfc committed:
Add issue number to comment

Add link in the comment explaining the workaround for getting workspace group
outputs from child algorithms to the issue that has been opened for this
on GitHub.

adriazalvarez reviewed (approved):
Thanks for taking the time to address my comments. 
I approved as the changes won't affect the test output.
